server/routes/room-routes.js

```diff
@@ -265,39 +282,84 @@ router.get(
+          removeMe_fetchRoomCreateEventId(matrixAccessToken, roomId),
+        ]);
+
+      // Without MSC3999, we currently only detect one kind of loop where the
+      // `m.room.create` has a timestamp that comes after the timestamp massaged events
+      // in the room. This is a common pattern for historical Gitter rooms where we
+      // created the room and then imported a bunch of messages at a time before the
+      // room was created.
+      //
+      // By nature of having an `timelineEndEventId`, we know we are already paginated
+      // past the `m.room.create` event which is always the first event in the room. So
+      // we can use that to detect the end of the room before we loop back around to the
+      // start of the room.
+      //
+      // XXX: Once we have MSC3999, we can remove this check in favor of that mechanism
+      // (TODO: Create issue to track this)
+      if (
+        dir === DIRECTION.forward &&
+        timelineEndEventId &&
+        eventIdForClosestEvent === roomCreateEventId
+      ) {
+        throw new StatusError(
+          404,
+          `/jump?dir=${dir}: We detected a loop back to the beginning of the room so we can assume ` +
+            `we hit the end of the room instead of doing a loop. We throw a 404 error here we hit ` +
+            `the normal 404 no more /messages error handling below`
+        );
+      }
 
       // Based on what we found was the closest, figure out the URL that will represent
       // the next chunk in the desired direction.
       // ==============================
       //
       // When jumping backwards, since a given room archive URL represents the end of
       // the day/time-period looking backward (scroll is also anchored to the bottom),
-      // we just need to get the user to the previous time-period.
+      // we just need to move the user to the time-period just prior the current one.
       //
       // We are trying to avoid sending the user to the same time period they were just
       // viewing. i.e, if they were visiting `/2020/01/02T16:00:00` (displays messages
       // backwards from that time up to the limit), which had more messages than we
       // could display in that day, jumping backwards from the earliest displayed event
-      // in the displayed range, say `T12:00:05` would still give us the same day
-      // `/2020/01/02` and we want to redirect them to previous chunk from that same
-      // day, like `/2020/01/02T12:00:00`
+      // in the displayed range (say that occured on `T12:00:25`) would still give us
+      // the same day `/2020/01/02` and we want to redirect them to previous chunk from
+      // that same day that still encompasses the closest message looking backwards,
+      // like `/2020/01/02T13:00:00`
       if (dir === DIRECTION.backward) {
+        // We choose `currentRangeStartTs` instead of `ts` (the jump point) because TODO: why?
+        // and we don't choose `currentRangeEndTs` because TODO: why?
```

Comment:
I feel like I can't justify this 🤔 but it does seem correct from tested behavior.

I think my reasoning for using `currentRangeEndTs` before was that we could keep paginating with day precision if we were showing more than a day for each page. But then this falls apart when you have a bunch of  messages from the same hour that fill up multiple pages.

Using `ts` makes various tests fail with various precision differences. Probably because of the `- 1`/`+ 1` stuff that we want to remove in the future anyway when MSC3999 lands.

But those don't seem like explanations for why (the truth kernel)

Reply:
I've left a link to this discussion thread so people can get some context. But I think I will opt to merging this PR without figuring this out. Will just have to contemplate this more and fill it out when an epiphany happens.